### PR TITLE
Add bundle-sqlcipher and bundle-ssl

### LINKED ISSUE #765: Bundled sqlcipher?
I've been experimenting with a fork of this repository where I enable building a bundled version of `sqlcipher`, optionally also bundling in an `openssl`/`libressl` libcrypto. Would you be interested in, or in principle open to a PR with this functionality?

Some related issues: #219, #558.

### PR DESCRIPTION
**Pull request by dubiousjim**
This reworks (and replaces) #771 on top of a more recent master. See that PR for some earlier discussion.

As mentioned there, we should confirm that Zetetic is ok with us re-distributing their amalgamated sources (which have to be built from sqlcipher/sqlcipher).

Would fix #765.

### CODE REVIEW

**Comment by dubiousjim** [comment]
Assuming this would be merged after #861

**Comment by dubiousjim** [comment]
Tested using this script:

```
#!/bin/sh -x
cargo fmt

printf -- '==== Clippy ====\n'
cargo clippy --all-targets --workspace --features "bundled"
cargo clippy --all-targets --workspace --features "bundled-full session buildtime_bindgen"

printf -- '==== Testing SQLITE ====\n'
cargo clean; cargo test
cargo clean; SQLITE3_LIB_DIR=/opt/local/lib cargo test
cargo clean; cargo test --all-targets --workspace --features="bundled"
cargo clean; cargo test --all-targets --workspace --features="bundled-full session buildtime_bindgen"

printf -- '==== Testing SQLCIPHER ====\n'
cargo clean; SQLCIPHER_LIB_DIR=/usr/local/lib cargo test --features="sqlcipher"
cargo clean; OPENSSL_LIB_DIR=/opt/local/lib OPENSSL_INCLUDE_DIR=/opt/local/include cargo test --features="bundled-sqlcipher"
cargo clean; OPENSSL_LIB_DIR=/opt/local/lib OPENSSL_INCLUDE_DIR=/opt/local/include cargo test --features="bundled-sqlcipher bundled-full session buildtime_bindgen"
cargo clean; cargo test --features="bundled-sqlcipher" # uses Apple crypto
cargo clean; cargo test --features="bundled-sqlcipher bundled-full session buildtime_bindgen"
cargo clean; cargo test --features="bundled-ssl"
cargo clean; cargo test --features="bundled-ssl bundled-full session buildtime_bindgen"
```

File: libsqlite3-sys/upgrade.sh [hunk 1]
```diff
@@ -1,28 +1,56 @@
...
 cargo update
 # Just to make sure there is only one bindgen.rs file in target dir
 find "$SCRIPT_DIR/../target" -type f -name bindgen.rs -exec rm {} \;
 env LIBSQLITE3_SYS_BUNDLING=1 cargo build --features "buildtime_bindgen" --no-default-features
-find "$SCRIPT_DIR/../target" -type f -name bindgen.rs -exec cp {} "$SQLITE3_LIB_DIR/bindgen_bundled_version.rs" \;
-# Sanity check
-cd "$SCRIPT_DIR/.." || { echo "fatal error"; exit 1; }
+find "$SCRIPT_DIR/../target" -type f -name bindgen.rs -exec mv {} "$SQLITE3_LIB_DIR/bindgen_bundled_version.rs" \;
+
+SQLCIPHER_VERSION="4.4.2"
```
Comment (gwenn) [inline libsqlite3-sys/upgrade.sh:30]: It would be nice if we can upgrade only sqlite **or** sqlcipher.

File: libsqlite3-sys/upgrade.sh [hunk 2]
```diff
@@ -1,28 +1,56 @@
...
+CUR_DIR=$(pwd -P)
 echo "$SCRIPT_DIR"
-cd "$SCRIPT_DIR" || { echo "fatal error"; exit 1; }
-export SQLITE3_LIB_DIR=$SCRIPT_DIR/sqlite3
+cd "$SCRIPT_DIR" || { echo "fatal error" >&2; exit 1; }
+cargo clean
+mkdir -p "$SCRIPT_DIR/../target" "$SCRIPT_DIR/sqlite3" "$SCRIPT_DIR/sqlcipher"
+export SQLITE3_LIB_DIR="$SCRIPT_DIR/sqlite3"
+export SQLITE3_INCLUDE_DIR="$SQLITE3_LIB_DIR"
+export SQLCIPHER_LIB_DIR="$SCRIPT_DIR/sqlcipher"
```
Comment (gwenn) [inline libsqlite3-sys/upgrade.sh:11]: You should add this new directory in `.gitattributes`:
```
libsqlite3-sys/sqlcipher/* linguist-vendored
```

File: libsqlite3-sys/upgrade.sh (hunk 1 above)
Comment (dubiousjim) [reply libsqlite3-sys/upgrade.sh:30]: Not hard to expand the script's api --- for example, with no arguments it'd "update" both (to the versions hard-coded in the script), or with an argument of "sqlite3" it'd only "update" that one, etc. The hardest part would be settling on the interface. But this is a script that one would run 2-5 times a year, and only takes about 5 minutes to "update" both, so I assumed that extra complexity wasn't worth it.

File: libsqlite3-sys/upgrade.sh (hunk 2 above)
Comment (dubiousjim) [reply libsqlite3-sys/upgrade.sh:11]: I don't yet understand this metadata, but have pushed a commit doing this.

Comment (dubiousjim) [reply libsqlite3-sys/upgrade.sh:11]: Seems this is a good starting point: <https://stackoverflow.com/questions/40659265/using-gitattributes-for-linguist-examples>

**Comment by christian-oudard** [comment]
We would really like to use this pull request in MobileCoin's wallet service. Is there anything we can do to help speed its adoption into the main library, or are we just going to have to use our fork of dubiousjim's code?

**Comment by thomcc** [comment]
Taking a look over it again, I'm largely fine with it. Truly sorry for the delays here.

@dubiousjim I'll r+ and merge if it gets rebased and (if necessary) ensures the version of sqlcipher that's vendored is the most up to date.

**Review by thomcc — APPROVED** [review]
This looks good to me now that I've updated things and added CI.

@gwenn any hesitations?

**Review by gwenn — APPROVED** [review]

**Comment by gwenn** [comment]
> This looks good to me now that I've updated things and added CI.
> 
> @gwenn any hesitations?

Nope

**Comment by thomcc** [comment]
I've renamed the `bundled-ssl` feature since it didn't sit right with me that it bundled sqlcipher, and turned on openssl-sys/vendored.

**Comment by thomcc** [comment]
Ah, hm, the README already says it's unix-only. I guess thats fine, although it's pretty close to working on windows at the same time. I guess that can be done as a follow-up.

**Pull request merged**